src/Framework/Constraint/Count.php

```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
             $this->expectedCount
         );
     }
+
+    private function getCountOfTraversable(Traversable $traversable): int
+    {
+        if ($this->traversableCounts === null) {
+            $this->traversableCounts = new \SplObjectStorage();
+        }
+
+        while ($traversable instanceof IteratorAggregate) {
+            $traversable = $traversable->getIterator();
+        }
+
+        if ($this->traversableCounts->contains($traversable)) {
+            return $this->traversableCounts[$traversable];
+        }
+
+        if ($traversable instanceof Iterator) {
+            return $this->countIteratorWithoutChangingPositionIfPossible($traversable);
+        }
+
+        return $this->getCountOfTraversableThatIsNotIteratorOrIteratorAggregate($traversable);
+    }
+
+    private function countIteratorWithoutChangingPositionIfPossible(Iterator $iterator): int
+    {
+        $key   = $iterator->key();
+        $count = $this->getCountOfIterator($iterator);
```

Comment:
Caution: original code uses custom iteration (without rewind) only for `Generator`, and `iterator_count` (which can rewind internally) for the rest, so I suspect this would change behavior (in edge cases). For example:

```php
$iterator = new \TestIterator([1, 2, 3]);
$iterator->next(); // this is the tricky bit

// which line below to uncomment, with original code? and with this PR?
//$n = 3;
//$n = 2;

Assert::assertCount($n, $iterator);
```

Reply:
This is tested by `testOnlyCountRemainingGeneratorElements`. I refactored that test to ascertain the original behavior and then made sure it stayed the same.

Reply:
I'm talking about *non*-`Generator` iterators:

```php
        $it = new \TestIterator([1, 2, 3]);
        $it->next();
        $this->assertTrue((new Count(3))->evaluate($it, '', true));
```

that was passing on master but is failing with this PR

Reply:
Oh right. The Iterator case was also not tested >_>

Reply:
Fixed now. Which allows getting rid of the ->detach()